File: internal/adapters/loader/actividades.go
// actividades.go — MA-T1b (paquete 2026-07-30-definicion-de-arnes, spec v2 §1 AUD-1..4):
// deriva el catálogo Arnes.Actividades + la faceta Box.Actividades desde el `arnes.yaml`
// que vive en la RAÍZ del arnés (misma base que arnes.l0.json). La derivación corre AL
// INDEXAR (patrón de derivarEdges) — jamás en el request, jamás al sello.
//
// Este lector NO reusa forja.ParseSemilla a propósito: (1) el grafo de imports prohíbe
// loader→forja (allow-list .go-arch-lint.yml); (2) las validaciones de siembra (territorios
// obligatorios, spines obligatorios) NO aplican acá — un arnes.yaml solo-con-tipos es legal
// para el Mapa aunque no alcance para sembrar. Subset propio, mínimo, con el vocabulario
// REAL de la semilla (`cierre`, id = clave del mapa — AUD-5).

package loader

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"

	"gopkg.in/yaml.v3"

	"github.com/alpacapurpura/arnesia/internal/domain"
)

// tipoPaqueteYAML es el subset de `gestion_trabajo.tipos_paquete.<id>` que el Mapa
// necesita (estados + cierre; jerarquia/wip_caps/appetite son de la siembra, no del Mapa).
type tipoPaqueteYAML struct {
	Estados []string `yaml:"estados"`
	Cierre  string   `yaml:"cierre"`
}

// pasoSpineYAML es el subset de un paso de `proceso.spines.<tipo>` que el Mapa necesita.
// `caja` es el campo nuevo de MA-T1a (AUD-1); ausente = «paso sin caja aún» (E13).
type pasoSpineYAML struct {
	Paso      string `yaml:"paso"`
	Rol       string `yaml:"rol"`
	Artefacto string `yaml:"artefacto"`
	Caja      string `yaml:"caja"`
}

// actividadesYAML es la forma cruda del subset. `tipos_paquete` se captura como yaml.Node
// para PRESERVAR el orden del documento (mismo requisito determinista que forja/parser.go:
// un map Go perdería el orden y el catálogo cambiaría entre corridas).
type actividadesYAML struct {
	GestionTrabajo struct {
		TiposPaquete yaml.Node `yaml:"tipos_paquete"`
	} `yaml:"gestion_trabajo"`
	Proceso struct {
		Spines map[string][]pasoSpineYAML `yaml:"spines"`
	} `yaml:"proceso"`
}

// leerActividades lee el `arnes.yaml` de la raíz del arnés y deriva el catálogo de
// actividades. Contrato de honestidad (mismo espíritu que leerManifiesto):
//   - archivo ausente → (nil, "", nil): arnés sin tipos declarados, legal (MA-L5/E8);
//   - archivo ilegible → (nil, aviso, nil): degradado VISIBLE, jamás aborta el grafo
//     (el grafo existe sin actividades; el aviso viaja en Info.Aviso);
//   - sin tipos NI spines → (nil, "", nil): nada que derivar, cero invento.
//
// Orden del catálogo: los tipos de `tipos_paquete` en ORDEN de documento; después los
// tipos que SOLO existen en `proceso.spines` (declarados a medias — visibles igual),
// en orden alfabético de clave para que el render sea determinista.
func leerActividades(dir string) ([]domain.Actividad, string, error) {
	ruta := filepath.Join(dir, "arnes.yaml")
	b, err := os.ReadFile(ruta) //nolint:gosec // G304: ruta bajo el dir del arnés que el caller eligió cargar.
	if errors.Is(err, os.ErrNotExist) {
		return nil, "", nil
	}
	if err != nil {
		return nil, "", fmt.Errorf("leer %s: %w", ruta, err)
	}

	var raw actividadesYAML
	if uerr := yaml.Unmarshal(b, &raw); uerr != nil {
		return nil, fmt.Sprintf("arnes.yaml ilegible — actividades no derivadas: %v", uerr), nil
	}

	var acts []domain.Actividad
	vistos := map[string]bool{}
	if werr := walkMappingActividades(&raw.GestionTrabajo.TiposPaquete, func(id string, val *yaml.Node) error {
		var tp tipoPaqueteYAML
		if derr := val.Decode(&tp); derr != nil {
			return fmt.Errorf("tipos_paquete.%s: %w", id, derr)
		}
		vistos[id] = true
		acts = append(acts, domain.Actividad{
			ID:      id,
			Estados: tp.Estados,
			Cierre:  tp.Cierre,
			Pasos:   pasosDe(raw.Proceso.Spines[id]),
		})
		return nil
	}); werr != nil {
		return nil, fmt.Sprintf("arnes.yaml tipos_paquete ilegible — actividades no derivadas: %v", werr), nil
	}

	// Spines de tipos NO declarados en tipos_paquete: declarados a medias, visibles igual
	// (honestidad — el Mapa los muestra; el gate de completitud D19 es de la forja).
	var huerfanos []string
	for tipo := range raw.Proceso.Spines {
		if !vistos[tipo] {
			huerfanos = append(huerfanos, tipo)
		}
	}
	sort.Strings(huerfanos)
	for _, tipo := range huerfanos {
		acts = append(acts, domain.Actividad{ID: tipo, Pasos: pasosDe(raw.Proceso.Spines[tipo])})
	}

	if len(acts) == 0 {
		return nil, "", nil
	}
	return acts, "", nil
}

// pasosDe traduce los pasos crudos del YAML al dominio (subset del Mapa).
func pasosDe(crudos []pasoSpineYAML) []domain.PasoActividad {
	if len(crudos) == 0 {
		return nil
	}
	pasos := make([]domain.PasoActividad, 0, len(crudos))
	for _, p := range crudos {
		pasos = append(pasos, domain.PasoActividad{Paso: p.Paso, Rol: p.Rol, Artefacto: p.Artefacto, Caja: p.Caja})
	}
	return pasos
}

// estamparFacetaActividades estampa en cada caja la faceta `actividades` (qué
// procedimientos la referencian vía pasos[].caja) — derivada, dedupeada, en el orden del
// catálogo. Una caja que ningún procedimiento referencia queda con la faceta vacía: ese
// silencio ES el dato (grupo `sin-actividad` del Mapa, E6). Un `caja:` que no matchea
// ningún nodo NO estampa nada ni inventa nodos — el hueco lo destapa el gate de
// conformance, no el loader.
func estamparFacetaActividades(acts []domain.Actividad, nodos []domain.Box) {
	porCaja := map[string][]string{}
	for _, a := range acts {
		for _, p := range a.Pasos {
			if p.Caja == "" {
				continue
			}
			if !contieneString(porCaja[p.Caja], a.ID) {
				porCaja[p.Caja] = append(porCaja[p.Caja], a.ID)
			}
		}
	}
	for i := range nodos {
		if ids, ok := porCaja[nodos[i].ID]; ok {
			nodos[i].Actividades = ids
		}
	}
}

// walkMappingActividades recorre un mapping yaml.v3 en orden de documento — copia local
// de forja.walkMapping (el grafo de imports prohíbe loader→forja; 12 líneas duplicadas
// valen menos que abrir ese boundary).
func walkMappingActividades(n *yaml.Node, fn func(key string, val *yaml.Node) error) error {
	if n == nil || n.Kind == 0 {
		return nil
	}
	if n.Kind != yaml.MappingNode {
		return fmt.Errorf("se esperaba un mapping, vino kind=%d", n.Kind)
	}
	for i := 0; i+1 < len(n.Content); i += 2 {
		if err := fn(n.Content[i].Value, n.Content[i+1]); err != nil {
			return err
		}
	}
	return nil
}

// contieneString reporta si xs contiene v (dedup de la faceta sin traer una dep).
func contieneString(xs []string, v string) bool {
	for _, x := range xs {
		if x == v {
			return true
		}
	}
	return false
}

File: web/src/entities/arnes/model/actividades.ts
// Derivaciones puras del catálogo multi-actividad (MA-T4/T5, spec 2026-07-30-definicion-de-arnes).
// El catálogo `arnes.actividades[]` y la faceta `box.actividades` llegan YA derivados por el
// loader al indexar (MA-T1b / CAP-151): acá NUNCA se re-deriva la faceta — solo se proyecta
// (qué cajas quedaron sin actividad, qué fases toca un procedimiento, qué salud agrega).
// Sin catálogo (arnés legacy, MA-L5/E8) todo devuelve vacío y el Mapa queda como hoy.
//
// La salud es HONESTA (spec §3): worst-of de los hallazgos FE ya existentes de sus cajas
// (gate:none crit · no-reconocido warn · paso sin caja / caja ausente warn) — la misma fuente
// del inspector, jamás una cifra tecleada ni un dato nuevo.

import { isCaja } from "./node-view"
import type { Actividad, Box, Graph } from "./types"

// SaludActividad — el agregado worst-of que colorea el dot del chip N0.
export type SaludActividad = "ok" | "warn" | "crit"

const peor = (a: SaludActividad, b: SaludActividad): SaludActividad => {
  if (a === "crit" || b === "crit") return "crit"
  if (a === "warn" || b === "warn") return "warn"
  return "ok"
}

// selectActividades — el catálogo del arnés, o [] (arnés sin tipos declarados, E1/E8).
export function selectActividades(g: Graph): Actividad[] {
  return g.arnes?.actividades ?? []
}

// cajasSinActividad — las cajas que NINGÚN procedimiento referencia (grupo `sin-actividad`,
// E6 — visible, insumo de poda §6). SOLO existe con catálogo declarado: sin tipos no hay
// pregunta que responder (MA-L5). La faceta viene del loader (box.actividades), no se re-deriva.
export function cajasSinActividad(g: Graph): Box[] {
  if (selectActividades(g).length === 0) return []
  return g.nodos.filter((n) => isCaja(n) && (n.actividades ?? []).length === 0)
}

// fasesDeActividad — el Set de fases que el procedimiento TOCA (las fases de las cajas de sus
// pasos). Alimenta E5: las fases fuera del set se atenúan enteras — se VE que el spike
// termina antes. Pasos sin caja (E13) no suman fase.
export function fasesDeActividad(g: Graph, actividad: Actividad): ReadonlySet<string> {
  const fases = new Set<string>()
  for (const p of actividad.pasos ?? []) {
    if (!p.caja) continue
    const n = g.nodos.find((x) => x.id === p.caja)
    if (n?.fase) fases.add(n.fase)
  }
  return fases
}

// saludDeActividad — worst-of derivado SOLO del grafo (sin dato nuevo, igual que el mockup):
// paso sin caja o caja ausente del grafo → warn · caja no-reconocido → warn ·
// caja con gate:none → crit (A4: el hueco de eval es un hallazgo, no un blanco).
export function saludDeActividad(g: Graph, actividad: Actividad): SaludActividad {
  let s: SaludActividad = "ok"
  for (const p of actividad.pasos ?? []) {
    if (!p.caja) {
      s = peor(s, "warn")
      continue
    }
    const n = g.nodos.find((x) => x.id === p.caja)
    if (!n) {
      s = peor(s, "warn")
      continue
    }
    if (n.clase === "no-reconocido") s = peor(s, "warn")
    if (n.contract?.gate?.tipo === "none") s = peor(s, "crit")
  }
  return s
}

// saludDeGrupo — el mismo worst-of para un grupo suelto de cajas (el chip `sin-actividad`).
export function saludDeGrupo(cajas: readonly Box[]): SaludActividad {
  let s: SaludActividad = "ok"
  for (const n of cajas) {
    if (n.contract?.gate?.tipo === "none") s = peor(s, "crit")
    else if (n.clase === "no-reconocido") s = peor(s, "warn")
  }
  return s
}
